### PR TITLE
MTD: Add module for remapping MTD sectors

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This MTD module allows for remapping multiple different regions on a single 
MTD device and present them as separate MTD devices. This is similar to     
partitions on a hard drive, although this system only allows hardcoded      
partitions and lacks a partition table.

The use case for this module is to be able to split a single MTD device, for
example a SPI NOR flash chip into multiple separate regions which all can   
contain their own content or file systems.

### Testing procedure

- run the provided test
- check the generated docs

### Issues/PRs references

None

### CODE REVIEW

File: tests/mtd_mapper/Makefile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+include ../Makefile.tests_common
+
+USEMODULE += mtd_mapper
+USEMODULE += embunit
+
+include $(RIOTBASE)/Makefile.include
+
```
Comment (kaspar030) [inline tests/mtd_mapper/Makefile]:
```suggestion
```

Comment (bergzand) [reply tests/mtd_mapper/Makefile]: Fixed!

File: tests/mtd_mapper/main.c [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    for (uint32_t i = 0; i < REGION_FLASH_SIZE; i += PAGE_SIZE) {
+        mtd_write(_dev_b, _buffer, i, PAGE_SIZE);
+    }
+
+    /* Check first region, should still be 0xAA */
+    for (uint32_t i = 0; i < REGION_FLASH_SIZE; i += PAGE_SIZE) {
+        mtd_read(_dev_a, _buffer, i, PAGE_SIZE);
+        _test_mem(_buffer, PAGE_SIZE, 0xAA);
+    }
+}
```
Comment (vincent-d) [inline tests/mtd_mapper/main.c]: you should maybe also check that the second region is properly written

File: drivers/include/mtd_mapper.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ * ```
+ *
+ * To define new regions with an existing MTD device the following is required:
+ *
+ * ```
+ * mtd_mapper_parent_t parent = {
+ *     .mtd = MTD_0,
+ *     .mutex = MUTEX_INIT,
+ *     .init = false,
+ * };
```
Comment (vincent-d) [inline drivers/include/mtd_mapper.h]: A macro might be handy to make sure `mutex` and `init` are properly initialized, like:

```
#define MTD_PARENT_INIT(_parent) \
{ \
    .mtd = _parent, \
    .mutex = MUTEX_INIT, \
    .init = false, \
}

File: tests/mtd_mapper/main.c (hunk 2 above)
Comment (bergzand) [reply tests/mtd_mapper/main.c]: Yup, fixed

File: drivers/include/mtd_mapper.h (hunk 3 above)
Comment (bergzand) [reply drivers/include/mtd_mapper.h]: Good one, added the macro and used it in both the test application and the example code

**Comment by kaspar030** [comment]
Please squash!

File: drivers/include/mtd_mapper.h [hunk 4]
```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (C) 2020  Koen Zandberg
```
Comment (kaspar030) [inline drivers/include/mtd_mapper.h]: Her and in the other files: there's an extra space. And usually we add an email address...

Comment (bergzand) [reply drivers/include/mtd_mapper.h]: > Her and in the other files: there's an extra space. And usually we add an email address...

Good catch, amended

**Comment by fjmolinas** [comment]
@kaspar030 does your squash request mean you also agree with the proposed api?

**Comment by bergzand** [comment]
@fjmolinas I've added a `@warning` to the module doc. Please let me know if this is what you had in mind.

**Comment by fjmolinas** [comment]
@bergzand thats OK with me, please squash!

**Comment by bergzand** [comment]
Squashed!

**Comment by bergzand** [comment]
Thanks!

File: drivers/include/mtd_mapper.h [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    bool init;          /**< Initialization flag */
+} mtd_mapper_parent_t;
+
+/**
+ * @brief MTD mapped region
+ */
+typedef struct {
+    mtd_dev_t mtd;                  /**< MTD context                         */
+    mtd_mapper_parent_t *parent;    /**< MTD mapper parent device            */
+    uint32_t offset;                /**< Offset address to start this region */
```
Comment (benpicco) [inline drivers/include/mtd_mapper.h:99]: I know I'm late, but can we give the offset in sectors?
The partitions have to be aligned to erase-sectors anyway and this way we can use in on media > 4GiB

**Pull request merged**